### PR TITLE
nvme: MAINT: Refactor telemetry-log command for build warning #1323

### PR DESCRIPTION
**Pull request by ikegami-t**
Signed-off-by: Tokunori Ikegami <ikegami.t@gmail.com>

### CODE REVIEW

**Comment by ikegami-t** [comment]
Sorry fixed the commit message subject error and also separated additionally the function.

File: common.h [hunk 1]
```diff
@@ -33,4 +33,6 @@ static inline uint64_t mmio_read64(void *addr)
 	return ((uint64_t) high << 32) | low;
 }
 
+#define fallthrough do {} while (0)
```
Comment (igaw) [inline common.h:36]: All relative modern [compilers](https://gcc.gnu.org/onlinedocs/gcc/Statement-Attributes.html) know about fallthrough attribute/statement. Hence you should test if the compiler supports it.

The Linux kernel used to do something like this
```
#if __has_attribute(__fallthrough__)
# define fallthrough                    __attribute__((__fallthrough__))
#else
# define fallthrough                    do {} while (0)  /* fallthrough */
#endif
```
but IIRC it was now dropped as all supported gcc/clang versions know about fallthrough

File: nvme.c [hunk 2]
```diff
@@ -407,20 +407,110 @@ static int get_ana_log(int argc, char **argv, struct command *cmd,
 	return nvme_status_to_errno(err, false);
 }
 
-static int get_telemetry_log(int argc, char **argv, struct command *cmd, struct plugin *plugin)
+static int get_telemetry_log_total_size_dalb4(int fd,
+					      struct nvme_telemetry_log *log,
+					      unsigned *total_size)
```
Comment (igaw) [inline nvme.c:412]: Could you also change from `unsigned` to `unsigned int` when touching this code?

File: nvme.c [hunk 3]
```diff
@@ -469,84 +559,19 @@ static int get_telemetry_log(int argc, char **argv, struct command *cmd, struct
...
-	}
-
-	err = write(output, (void *)log, total_size);
-	if (err != total_size) {
-		fprintf(stderr, "Failed to flush all data to file!\n");
-		goto close_output;
+	err = get_telemetry_log_total_size(fd, cfg.data_area, log, &total_size);
+	if (total_size) {
+		err = write(output, (void *)log, total_size);
+		if (err != total_size)
```
Comment (igaw) [inline nvme.c:573]: Since err and total_size have two different types, it's probably better to do:
`if (err < 0 || err != total_size)`

**Review by igaw — CHANGES_REQUESTED** [review]

File: nvme.c (hunk 2 above)
Comment (ikegami-t) [reply nvme.c]: Fixed as mentioned.

File: nvme.c (hunk 3 above)
Comment (ikegami-t) [reply nvme.c]: Fixed this also.

**Comment by ikegami-t** [comment]
Thasnk for your comments. Fixed the code as you mentioned so please review again.

**Comment by ikegami-t** [comment]
Additionally changed the code as below.
1. Changed "unsigned" still remained to "unsigned  int".
2. Changed "long long unsigned int" to "unsigned long long".

**Comment by igaw** [comment]
Looks good. I just splited the type changes out of the initial patch.

**Comment by igaw** [comment]
Thanks!

**Pull request merged**